[mlir] Prepare convert-gpu-to-spirv for OpenCL support

silee2 opened the pull request:
This includes a couple of changes to pass behavior for OpenCL kernels. Vulkan shaders are not impacted by the changes.

1. SPIRV module is placed inside GPU module. This change is required for gpu-module-to-binary to work correctly as it expects kernel function to be inside the GPU module.
2. A dummy func.func with same kernel name as gpu.func is created. GPU compilation pipeline defers lowering of gpu launch kernel op. Since spirv.func is not directly tied to gpu launch kernel, a dummy func.func is required to avoid legalization issues.
3. Use correct mapping when mapping MemRef memory space to SPIR-V storage class for OpenCL kernels.

silee2 committed:
[MLIR] Update convert-gpu-to-spirv pass to prepare using GPU compilation pipeline for OpenCL kernels.

This includes a couple of changes to pass behavior for OpenCL kernels.
Vulkan shaders are not impacted by the changes.

1. SPIRV module is placed inside GPU module. This change is required for
gpu-module-to-binary to work correctly as it expects kernel function to be
inside the GPU module.
2. A dummy func.func with same kernel name as gpu.func is created.
GPU compilation pipeline defers lowering of gpu launch kernel op.
Since spirv.func is not directly tied to gpu launch kernel,
a dummy func.func is required to avoid legalization issues.
3. Use correct mapping when mapping MemRef memory space to SPIR-V storage class for OpenCL kernels.

silee2 commented:
Replacing #69941 since a new pass flag was not required.

silee2 commented:
This PR is one of the components for breaking down #65539 into smaller PRs and use new GPU compilation pipeline.

kuhar commented on mlir/include/mlir/Conversion/Passes.td:
```diff
@@ -578,7 +578,10 @@ def ConvertGPUToSPIRV : Pass<"convert-gpu-to-spirv", "ModuleOp"> {
   let constructor = "mlir::createConvertGPUToSPIRVPass()";
-  let dependentDialects = ["spirv::SPIRVDialect"];
+  let dependentDialects = [
+    "spirv::SPIRVDialect",
+    "func::FuncDialect",
```
nit: IIRC we try to keep these sorted alphabetically
```suggestion
    "func::FuncDialect",
    "spirv::SPIRVDialect",
```

silee2 replied:
Done.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,63 @@ void GPUToSPIRVPass::runOnOperation() {
   SmallVector<Operation *, 1> gpuModules;
   OpBuilder builder(context);
+
+  auto getTargetEnvFromGPUModuleOp = [=](gpu::GPUModuleOp moduleOp) {
```
Can you list the captured variables explicitly?

silee2 replied:
Done.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,63 @@ void GPUToSPIRVPass::runOnOperation() {
+    // This works fine for Vulkan shader that has a dedicated runner.
+    // But OpenCL kernel needs SPIRV module placed inside original GPU module as
+    // OpenCL uses GPU compilation pipeline.
+    auto targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
```
nit: I realize this comes from existing code elsewhere in this file, but could you write the full type here instead of `auto`? It's not obvious from the RHS, which is the general criterium for `auto` in MLIR.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,63 @@ void GPUToSPIRVPass::runOnOperation() {
+        spirv::getMemoryModel(targetEnv);
+    if (failed(memoryModel))
+      return signalPassFailure();
+    (memoryModel == spirv::MemoryModel::OpenCL)
+        ? builder.setInsertionPoint(moduleOp.getBody(),
+                                    moduleOp.getBody()->begin())
+        : builder.setInsertionPoint(moduleOp.getOperation());
```
Use `if` `else` for readability.

silee2 replied:
Done.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,63 @@ void GPUToSPIRVPass::runOnOperation() {
   // Run conversion for each module independently as they can have different
   // TargetEnv attributes.
   for (Operation *gpuModule : gpuModules) {
+    auto targetAttr = spirv::lookupTargetEnvOrDefault(gpuModule);
```
also here

silee2 replied:
Done.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -108,6 +143,29 @@ void GPUToSPIRVPass::runOnOperation() {
     if (failed(applyFullConversion(gpuModule, *target, std::move(patterns))))
       return signalPassFailure();
   }
+  // In case of OpenCL, gpu.func in original gpu.module needs to replaced with
```
nit: add a newline before a large block of code like this

silee2 replied:
Done.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -108,6 +143,29 @@ void GPUToSPIRVPass::runOnOperation() {
+  // an empty func.func with same arguments as gpu.func. And it also needs
+  // gpu.kernel attribute set.
+  module.walk([&](gpu::GPUModuleOp moduleOp) {
+    auto targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
```
also here

silee2 replied:
Done.

kuhar reviewed (commented):
Left some stylistic suggestions. @antiagainst, could you also take a look?

silee2 committed:
Address reviewer comments.
Sort alphabetically.
Use if-else instead of ternary expression.
List lambda captures explicitly.
Replace auto with actually type.
Add newline before big block.

silee2 committed:
Run clang formatter.

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -76,22 +76,24 @@ void GPUToSPIRVPass::runOnOperation() {
     // But OpenCL kernel needs SPIRV module placed inside original GPU module as
     // OpenCL uses GPU compilation pipeline.
-    auto targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
+    const mlir::spirv::TargetEnv& targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
```
```suggestion
    const spirv::TargetEnv& targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
```

silee2 replied:
Removed mlir::

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -76,22 +76,24 @@ void GPUToSPIRVPass::runOnOperation() {
   // TargetEnv attributes.
   for (Operation *gpuModule : gpuModules) {
-    auto targetAttr = spirv::lookupTargetEnvOrDefault(gpuModule);
+    mlir::spirv::TargetEnvAttr targetAttr = spirv::lookupTargetEnvOrDefault(gpuModule);
```
```suggestion
    spirv::TargetEnvAttr targetAttr = spirv::lookupTargetEnvOrDefault(gpuModule);
```

silee2 replied:
Removed mlir::

kuhar commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -143,11 +145,12 @@ void GPUToSPIRVPass::runOnOperation() {
   // gpu.kernel attribute set.
   module.walk([&](gpu::GPUModuleOp moduleOp) {
-    auto targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
+    const mlir::spirv::TargetEnv& targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
```
```suggestion
    const spirv::TargetEnv& targetEnv = getTargetEnvFromGPUModuleOp(moduleOp);
```

silee2 replied:
Removed mlir::

kuhar reviewed (approved):
LGTM % nits, but let's give @antiagainst a chance to take a look.

antiagainst commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,67 @@ void GPUToSPIRVPass::runOnOperation() {
+    SPIRVConversionOptions options;
+    options.use64bitIndex = this->use64bitIndex;
+    SPIRVTypeConverter typeConverter(targetAttr, options);
+    const spirv::TargetEnv &targetEnv = typeConverter.getTargetEnv();
```
You can create a `spirv::TargetEnv` directly from a `spirv::TargetEnvAttr`; don't need to go through the type converter.

silee2 replied:
Thanks! That simplifies the code a lot.

antiagainst commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,67 @@ void GPUToSPIRVPass::runOnOperation() {
     // launch op still needs the original GPU kernel module.
-    builder.setInsertionPoint(moduleOp.getOperation());
+    // SPIRV module insertion point by is after original GPU module.
+    // This works fine for Vulkan shader that has a dedicated runner.
```
I'd suggest to change the comment to something like "For Vulkan Shader capabilities, we insert the newly converted SPIR-V module right after the original GPU module, as that's the expectation of the in-tree Vulkan runner. For OpenCL Kernel capabilities, we insert the newly converted SPIR-V module inside the original GPU module, as that's the expecation of the normal GPU compilation pipeline."

silee2 replied:
Thank for the suggestion. Updated comment.

antiagainst commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,67 @@ void GPUToSPIRVPass::runOnOperation() {
   // Run conversion for each module independently as they can have different
   // TargetEnv attributes.
   for (Operation *gpuModule : gpuModules) {
+    mlir::spirv::TargetEnvAttr targetAttr =
```
Can we use the lambda function in the above? This is duplicating the logic there.

silee2 replied:
Now the lambda returns boolean and cannot be used. Also that line is from the original code. I just moved it up a bit since I needed access to targetAttr before memory space mapping.

antiagainst commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -108,6 +147,31 @@ void GPUToSPIRVPass::runOnOperation() {
       return signalPassFailure();
   }
+
+  // In case of OpenCL, gpu.func in original gpu.module needs to replaced with
```
For OpenCL, the gpu.func op in the original gpu.module op ...

silee2 replied:
Updated comment.

antiagainst commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,67 @@ void GPUToSPIRVPass::runOnOperation() {
+    // OpenCL uses GPU compilation pipeline.
+    const mlir::spirv::TargetEnv &targetEnv =
+        getTargetEnvFromGPUModuleOp(moduleOp);
+    FailureOr<spirv::MemoryModel> memoryModel =
```
I think instead of checking the memory model, we want to check whether `targetEnv.allows(spirv::Capability::Kernel)`. 

Actually maybe we put it directly in the above lambda and rename it as something like `targetEnvSupportsKernelCapability`.

silee2 replied:
That is a great idea. Thanks! Updated the lambda.

antiagainst commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,67 @@ void GPUToSPIRVPass::runOnOperation() {
-    builder.setInsertionPoint(moduleOp.getOperation());
+    // SPIRV module insertion point by is after original GPU module.
+    // This works fine for Vulkan shader that has a dedicated runner.
+    // But OpenCL kernel needs SPIRV module placed inside original GPU module as
```
Is it possible to unify these two paths actually?

silee2 replied:
I don't know. My goal was to make no change to existing Vulkan workflow.

antiagainst reviewed (changes requested):
Thanks for adding support for this! Overall LGTM; just a few inline comments.

silee2 committed:
Address reviewers suggestions.

silee2 committed:
Merge branch 'main' into opencl_memspace

silee2 committed:
Check for spirv::MemoryModel only if mapping MemRef memory space to SPIR-V storage class.

silee2 committed:
Remove redundant namespace resolution.

silee2 commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,67 @@ void GPUToSPIRVPass::runOnOperation() {
     // Clone each GPU kernel module for conversion, given that the GPU
     // launch op still needs the original GPU kernel module.
-    builder.setInsertionPoint(moduleOp.getOperation());
+    // SPIRV module insertion point by is after original GPU module.
```
Done.

silee2 commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,67 @@ void GPUToSPIRVPass::runOnOperation() {
+  auto getTargetEnvFromGPUModuleOp = [*this](gpu::GPUModuleOp moduleOp) {
+    Operation *gpuModule = moduleOp.getOperation();
+    auto targetAttr = spirv::lookupTargetEnvOrDefault(gpuModule);
+    std::unique_ptr<ConversionTarget> target =
```
Removed usage of target.

antiagainst commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -54,22 +55,52 @@ void GPUToSPIRVPass::runOnOperation() {
           spirv::getMemorySpaceToStorageClassTarget(*context);
       spirv::MemorySpaceToStorageClassMap memorySpaceMap =
-          spirv::mapMemorySpaceToVulkanStorageClass;
+          (memoryModel == spirv::MemoryModel::OpenCL)
```
Here we can also use the above `targetEnvSupportsKernelCapability` no? Something like:

```c++
spirv::MemorySpaceToStorageClassMap memorySpaceMap =
  targetEnvSupportsKernelCapability(gpuModule) ? ...
```

silee2 replied:
Done.

antiagainst reviewed (changes requested):
Nice. Just one final request.

silee2 committed:
Another reviewer suggestion.

silee2 committed:
Merge branch 'main' into opencl_memspace

silee2 commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRVPass.cpp:
```diff
@@ -108,6 +138,25 @@ void GPUToSPIRVPass::runOnOperation() {
       return signalPassFailure();
   }
+
+  // For OpenCL, the gpu.func op in the original gpu.module op needs to be
```
Keeping the original gpu.func causes legality check error later in the gpu compile pipeline. 
If target attr is set for a gpu.module, gpu-to-llvm pass doesn't lower gpu.launch_func 
Instead, it is replaced with another gpu.launch_func that has lowered argument types (llvm ptrs).
If a gpu.func remains as an input to gpu-to-llvm pass, there is an argument mismatch between the new gpu.launch_func and the gpu.func. And a error is fired.
The reason for putting a dummy func.func here is to work around that check.
For func types other than gpu.func, argument types are not checked against gpu.launch_func.
But a func.func is still need as there will be a symbol check.

jungpark-mlir replied:
I think the problem is, the lack of spirv support in gpu dialect. For example, gpu.func needs to be able to wrap spirv.func so gpu-to-llvm pass (for the host code) can properly handle the relation between gpu.launch_func and spirv.func.
Basically, using dummy `func.func` looks little hacky and it'd be also nice if the divergence between Vulkan/OpenCL IR structures could be avoided. However, considering the progress of this commit, we can discuss this later for the future enhancement. 
Really appreciate this work and look forward to seeing it merged.

antiagainst reviewed (approved):
LGTM. Thanks!

antiagainst commented:
I've updated the tile and commit message slightly to make it look nicer once land as a git commit. Thanks!

silee2 commented:
@antiagainst @joker-eph Can someone merge this PR?

antiagainst commented:
> @antiagainst @joker-eph Can someone merge this PR? Thanks in advance.

Done. Thanks for the contribution again!
